StrongInject.Tests.Unit/OwnedInjectionTests.cs

```diff
@@ -0,0 +1,1275 @@
+        global::B b_0_1;
+        global::StrongInject.Owned<global::C> owned_0_2;
+        global::D d_0_4;
+        global::A a_0_0;
+        b_0_1 = new global::B();
+        try
+        {
+            global::StrongInject.Owned<global::C> CreateOwnedC_3()
+            {
+                global::C c_0_0;
+                c_0_0 = new global::C();
+                return new global::StrongInject.Owned<global::C>(c_0_0, () =>
+                {
+                });
+            }
+
+            owned_0_2 = CreateOwnedC_3();
+            d_0_4 = new global::D();
+            try
+            {
+                a_0_0 = new global::A(b: b_0_1, c: owned_0_2, d: d_0_4);
+            }
+            catch
+            {
+                ((global::System.IDisposable)d_0_4).Dispose();
+                throw;
+            }
+        }
+        catch
+        {
+            ((global::System.IDisposable)b_0_1).Dispose();
+            throw;
+        }
+
+        TResult result;
+        try
+        {
+            result = func(a_0_0, param);
+        }
+        finally
+        {
+            ((global::System.IDisposable)d_0_4).Dispose();
+            ((global::System.IDisposable)b_0_1).Dispose();
+        }
+
+        return result;
+    }
+
+    global::StrongInject.Owned<global::A> global::StrongInject.IContainer<global::A>.Resolve()
+    {
+        if (Disposed)
+            throw new global::System.ObjectDisposedException(nameof(Container));
+        global::B b_0_1;
+        global::StrongInject.Owned<global::C> owned_0_2;
+        global::D d_0_4;
+        global::A a_0_0;
+        b_0_1 = new global::B();
+        try
+        {
+            global::StrongInject.Owned<global::C> CreateOwnedC_3()
+            {
+                global::C c_0_0;
+                c_0_0 = new global::C();
+                return new global::StrongInject.Owned<global::C>(c_0_0, () =>
+                {
+                });
+            }
+
+            owned_0_2 = CreateOwnedC_3();
+            d_0_4 = new global::D();
+            try
+            {
+                a_0_0 = new global::A(b: b_0_1, c: owned_0_2, d: d_0_4);
+            }
+            catch
+            {
+                ((global::System.IDisposable)d_0_4).Dispose();
+                throw;
+            }
+        }
+        catch
+        {
+            ((global::System.IDisposable)b_0_1).Dispose();
+            throw;
+        }
+
+        return new global::StrongInject.Owned<global::A>(a_0_0, () =>
+        {
+            ((global::System.IDisposable)d_0_4).Dispose();
+            ((global::System.IDisposable)b_0_1).Dispose();
+        });
+    }
+}");
+        }
+
+        [Fact]
+        public void InjectParameterFuncOfAsyncOwnedOfDisposableDependency()
+        {
+            string userSource = @"
+using System;
+using System.Threading.Tasks;
+using StrongInject;
+
+[Register(typeof(A)), Register(typeof(B))]
+public partial class Container : IAsyncContainer<A> { }
```

Comment:
This shouldn't require an`IAsyncContainer` - can you test that we don't get an error if we use a normal `IContainer` please.